processing/src/main/java/io/druid/segment/VirtualColumns.java

```diff
@@ -19,62 +19,250 @@
+      }
+
+      if (withDotSupport.containsKey(vc.getOutputName()) || withoutDotSupport.containsKey(vc.getOutputName())) {
+        throw new IAE("Duplicate virtualColumn name[%s]", vc.getOutputName());
+      }
+
       if (vc.usesDotNotation()) {
         withDotSupport.put(vc.getOutputName(), vc);
       } else {
         withoutDotSupport.put(vc.getOutputName(), vc);
       }
     }
-    return new VirtualColumns(withDotSupport, withoutDotSupport);
+    return new VirtualColumns(ImmutableList.copyOf(virtualColumns), withDotSupport, withoutDotSupport);
   }
 
-  public VirtualColumns(Map<String, VirtualColumn> withDotSupport, Map<String, VirtualColumn> withoutDotSupport)
+  private VirtualColumns(
+      List<VirtualColumn> virtualColumns,
+      Map<String, VirtualColumn> withDotSupport,
+      Map<String, VirtualColumn> withoutDotSupport
+  )
   {
+    this.virtualColumns = virtualColumns;
     this.withDotSupport = withDotSupport;
     this.withoutDotSupport = withoutDotSupport;
+
+    for (VirtualColumn virtualColumn : virtualColumns) {
+      detectCycles(virtualColumn, null);
+    }
   }
 
+  // For equals, hashCode, toString, and serialization:
+  private final List<VirtualColumn> virtualColumns;
+
+  // For getVirtualColumn:
   private final Map<String, VirtualColumn> withDotSupport;
   private final Map<String, VirtualColumn> withoutDotSupport;
 
-  public VirtualColumn getVirtualColumn(String dimension)
+  public boolean exists(String columnName)
+  {
+    return getVirtualColumn(columnName) != null;
+  }
+
+  public VirtualColumn getVirtualColumn(String columnName)
   {
-    VirtualColumn vc = withoutDotSupport.get(dimension);
+    final VirtualColumn vc = withoutDotSupport.get(columnName);
     if (vc != null) {
       return vc;
     }
-    for (int index = dimension.indexOf('.'); index >= 0; index = dimension.indexOf('.', index + 1)) {
-      vc = withDotSupport.get(dimension.substring(0, index));
-      if (vc != null) {
-        return vc;
-      }
+    final String baseColumnName = splitColumnName(columnName).lhs;
+    return withDotSupport.get(baseColumnName);
+  }
+
+  public ObjectColumnSelector makeObjectColumnSelector(String columnName, ColumnSelectorFactory factory)
+  {
+    final VirtualColumn virtualColumn = getVirtualColumn(columnName);
+    if (virtualColumn == null) {
+      return null;
+    } else {
+      return Preconditions.checkNotNull(
+          virtualColumn.makeObjectColumnSelector(columnName, factory),
+          "VirtualColumn[%s] returned a null ObjectColumnSelector for columnName[%s]",
+          virtualColumn.getOutputName(),
+          columnName
+      );
+    }
+  }
+
+  public DimensionSelector makeDimensionSelector(DimensionSpec dimensionSpec, ColumnSelectorFactory factory)
+  {
+    final VirtualColumn virtualColumn = getVirtualColumn(dimensionSpec.getDimension());
+    if (virtualColumn == null) {
+      return dimensionSpec.decorate(NullDimensionSelector.instance());
```

Comment:
I think this is an extension of an existing convention where reading from a non-existent column is treated as reading from a column full of nulls, e.g.:

```
final Column columnDesc = index.getColumn(dimension);
if (columnDesc == null) {
  return NULL_DIMENSION_SELECTOR;
}
```

from `QueryableIndexStorageAdapter`